File: uniquePath/uniquePaths_exhaustive.py
'''
Time Complexity : 0(2^(mn))
Space Complexity : Recursion Stack 0(m+n)
On LeetCode: Time Limit Exceed
'''

class Solution:
    
    def __init__(self):
        self.grid = None
    
    def exhaustiveDfs(self,m,n):
        # base-case
        if (m < 0 or m == len(self.grid)) or (n<0 or n==len(self.grid[0])):
            return
        
        # logic-case
        # increase the value of the position
        if self.grid[m][n] == None:
            self.grid[m][n] = 1
        else:
            self.grid[m][n] += 1
        
        # dfs on rhs
        self.exhaustiveDfs(m,n+1)
        
        # dfs bottom
        self.exhaustiveDfs(m+1,n)
    
    def uniquePaths(self, m: int, n: int) -> int:
        
        # initlize a grid
        self.grid = [[None for col in range(0,n)] for row in range(0,m)]
        
        # peform exhaustive dfs
        self.exhaustiveDfs(0,0)
        
        # return the result
        return self.grid[-1][-1]

File: unique-paths-exhaustive.cpp
// Time Complexity : O(m^2 x n^2)
// Space Complexity : O(m^2 x n^2)
// Did this code successfully run on Leetcode : NO (Time Limit Exceeded)

#include <iostream>

using namespace std;

class Solution {
public:
    int uniquePaths(int m, int n) {
        return helper(m, n, 0, 0);
    }
private:
    int helper(int m, int n, int i, int j){
        //base
        if(i == m || j == n) return 0;
        if(i == m - 1 || j == n - 1) return 1;
        //logic
        //right
        int case1 = helper(m, n, i, j+1);
        //left
        int case2 = helper(m, n, i+1, j);
        return case1 + case2;
    }
};